### ISSUE TITLE
Build from local Dockerfile

### ISSUE THREAD

**Issue by JimNero009** [testcontainers/testcontainers-node#3]
Hey,

Just a quick note asking if building images from a local Dockerfile was on your roadmap? From what I can see, that is not currently supported. To be clear, what I mean is creating a GenericContainer from a Dockerfile that you keep within the repository itself.

**Comment by cristianrgreco** [comment]
Hey @JimNero009 

I had a similar requirement last week but I opted to build and push the image to DockerHub instead.
I didn't think it was the responsibility of testcontainers to build the image as the API could get quite complex (build args, etc).

**Comment by JimNero009** [comment]
Hi Christian,
Thank you for your answer. The reason I ask is because I see that the Java library has support for this, and the Python library has similar support (in the context of a docker-compose file, in any case) so it seemed a logical extension to include it here. I personally think it would be an incredibly powerful feature that would allow node developers everywhere to properly test artefacts they are producing _before_ pushing it anywhere, speeding up iteration time.

Just food for thought :)

**Comment by cristianrgreco** [comment]
Hey @JimNero009 

I've started work on this feature. What do you think about the following API:

```javascript
const { GenericContainer } = require("testcontainers");

(async () => {
  const context = "/src";
  const src = ["Dockerfile", "index.js"];
  const imageName = "my-custom-image";
  const tag = "1.0.0";
  const container = await GenericContainer.fromDockerfile(imageName, tag, context, src);
  
  const startedContainer = await container
      .withEnv("KEY", "VALUE")
      .withExposedPorts(8080)
      .start();

  await startedContainer.stop();
})();
```

**Comment by JimNero009** [comment]
Hey @cristianrgreco -- sorry for the tardiness of my reply.

I think we're looking pretty good with that! A couple of suggestions:
1. Is it really necessary to have an image name and tag seperated like that? Since this is a container being built from a local Dockerfile, I think an optional parameter that specifies an image name for later reference is sufficient here. In the use cases I'm thinking of, knowing the name of the container I've created is irrelevant and should be abstracted away from me.
2. If I'm understanding correctly, you are explicitly passing in a list of file names that the Docker deamon needs to build the image from the Dockerfile? I quite like that, but seeing as there is a strong convention for dockerfiles to be called 'Dockerfile;, there is maybe something we can do around defaults here.

For my understanding, where is the path passed into 'context' relative to? I can imagine it making sense to be relative to a project root.

Nice work, and super glad you've decided to look at this feature :)

**Comment by cristianrgreco** [comment]
Hey @JimNero009,

Thanks for the feedback, both great points.

In terms of the first point, if the name of the container is abstracted away, does it make sense to keep the image once the application has terminated? I see that the Java port of testcontainers will auto-delete the image by default on shutdown, wondering if we should do the same?

For the second point, after having gotten feedback from members of the Java port of testcontainers, I think I'll be changing this so you just provide a context path, and we use all the files in that directory. I think it makes sense because the Dockerfile already specifies what should be copied into the image.

The path can be either relative or absolute, I was thinking it could be used like this:
```javascript
const context = "/tmp";  // absolute path in the FS
const context = path.resolve(__dirname, "dir") // relative from the current directory of the node module
```

**Comment by ebenson6295** [comment]
Hello! First of all, your implementation of `testcontainers` is amazing. Thank you so much for all of your work. I just wanted to ask about this feature (building from local Dockerfile) and see if you have an ETA for when it will be completed. It seems like it will be super awesome.

**Comment by cristianrgreco** [comment]
Hey @JimNero009, @ebenson6295,

Thanks for the positive feedback!

This feature is now available in version `1.1.19`.
Example usage from the `README`:

```javascript
const path = require('path');
const { GenericContainer } = require("testcontainers");

(async () => {
  const buildContext = path.resolve(__dirname, "my-dir");
  
  const container = await GenericContainer.fromDockerfile(buildContext);
  
  const startedContainer = await container
      .withEnv("KEY", "VALUE")
      .withExposedPorts(8080)
      .start();

  await startedContainer.stop();
})();
```

**Comment by eli-lim** [comment]
Hey, thanks for the awesome package! ❤️

I'm using the `GenericContainer.fromDockerFile` API, trying to give the built image a `name` and `tag` so I can avoid rebuilding the image every test run.

Correct me if I'm wrong, but I quickly looked through the codebase and I don't think it's possible atm?
